src/mercury_engine_data_structures/formats/bmsbk.py

```diff
@@ -41,37 +43,108 @@ def _rebuild_types(ctx: Container) -> int:
+    BABY = "baby"
+    SCREW_ATTACK = "screw_attack"
+    WEIGHT = "weight"
+
+
+BlockTypeConstruct = EnumAdapter(BlockType, StrId)
+
 BlockGroup = Struct(
     "_num_blocks" / Rebuild(Int32ul, _rebuild_blocks),
     "_num_types" / Rebuild(Int32ul, _rebuild_types),
     "is_enabled" / Flag, # always true?
     "types" / Array(lambda this: this._num_types, Struct(
-        "block_type" / StrId,
+        "block_type" / BlockTypeConstruct,
         "blocks" / make_vector(Block),
     )),
 )  # fmt: skip
 
 BMSBK = Struct(
-    "magic" / Const(b"MSBK"),
+    "_magic" / Const(b"MSBK"),
     "version" / VersionAdapter("1.10.0"),
     "block_groups" / make_vector(BlockGroup),
-    "collision_cameras" / make_vector(Struct(
-        "name" / StrId,
-        "entries" / make_vector(Int32ul),
-    )),
+    "collision_cameras" / make_dict(make_vector(Int32ul)),
     construct.Terminated,
 )  # fmt: skip
 
 
-class BlockType(Enum):
-    POWER_BEAM = "power_beam"
-    BOMB = "bomb"
-    MISSILE = "missile"
-    SUPER_MISSILE = "super_missile"
-    POWER_BOMB = "power_bomb"
-    BABY = "baby"
-    SCREW_ATTACK = "screw_attack"
-    WEIGHT = "weight"
+class BlockData:
+    def __init__(self, raw: Container) -> None:
+        self._raw = raw
+
+    @property
+    def position(self) -> Vec3:
+        return self._raw.position
+
+    @position.setter
+    def position(self, value: Vec3) -> None:
+        self._raw.position = value
+
+    @property
+    def respawn_time(self) -> float:
+        return self._raw.respawn_time
+
+    @respawn_time.setter
+    def respawn_time(self, value: float) -> None:
+        self._raw.respawn_time = value
+
+    @property
+    def model_name(self) -> str:
+        return self._raw.model_name
+
+    @model_name.setter
+    def model_name(self, value: str) -> None:
+        self._raw.model_name = value
+
+    @property
+    def vignette_name(self) -> str:
+        return self._raw.vignette_name
+
+    @vignette_name.setter
+    def vignette_name(self, value: str) -> None:
+        self._raw.vignette_name = value
+
+
+class BlockGroupData:
+    def __init__(self, raw: Container):
+        self._raw = raw
+
+    @property
+    def block_type(self) -> BlockType:
+        return self._raw.types[0].block_type
+
+    @block_type.setter
+    def block_type(self, block_type: BlockType) -> None:
+        self._raw.types[0].block_type = block_type
+
+    def get_block(self, block_idx: int) -> BlockData:
+        return BlockData(self._raw.types[0].blocks[block_idx])
+
+    def add_block(
+        self, position: Vec3, respawn_time: float, model_name: str, vignette_name: str, unk2: int = 0
+    ) -> Container:
+        new_block = Container(
+            {
+                "position": position,
+                "unk2": unk2,
+                "unk3": 0,
+                "respawn_time": respawn_time,
+                "model_name": model_name,
+                "vignette_name": vignette_name,
+            }
+        )
```

Comment:
this should probably be a classmethod on `BlockData`:
```py
class BlockData:
    @classmethod
    def create(cls, position: Vec3, respawn_time: float, model_name: str, vignette_name: str, unk2: int = 0) -> typing.Self:
        return cls(Container({
            "position": position,
            "unk2": unk2,
            "unk3": 0,
            "respawn_time": respawn_time,
            "model_name": model_name,
            "vignette_name": vignette_name,
        }))
```

Reply:
Ah I've never used classmethod before. Seems cleaner.